### ISSUE TITLE
custom operators by reference in tf2.2.0

### ISSUE THREAD

**Issue by scarrazza** [tensorflow/tensorflow#40006]
**System information**
- TensorFlow version (you are using): 2.2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Following the [custom operator guide](https://www.tensorflow.org/guide/create_op) we learn how to build a custom operator.

However, the guide does not explain how to build custom operators when the input and output are passed by reference. In fact, following the [AssignOp](https://github.com/tensorflow/tensorflow/blob/7817237d5852c7778d3ba03c40f139e6f2c37a76/tensorflow/core/kernels/assign_op.h) as a naive example, we get errors concerning the impossibility to run it on eager mode and when using `@tf.function` / disabling eager that the input tensors must be mutable, even if the object we are passing is already a `tf.Variable`.

Searching for the `tf.Variable.assign` function in v2.2.0 we realize that its c++ operator is [AssignVariableOp](https://github.com/tensorflow/tensorflow/blob/7769f52b9f6433c4e97ae42ba2060034e357a071/tensorflow/core/kernels/resource_variable_ops.cc) which is interfaced in `gen_resource_variable_ops` (generated during compilation) and it performs the `tf.Variable` updates through a resource object mechanism.

As far as I understand, there is no simple side load mechanism for custom operators by reference without modifying the `gen_resource_variable_ops` / `resource_variable_ops.py`, thus I am wondering if there are less intrusive alternatives to this approach and the respective documentation.

**Will this change the current api? How?**

Probably yes.

**Who will benefit with this feature?**

Projects that require custom operators based on mutable objects.

**Comment by mdanatg** [comment]
cc @saxenasaurabh @allenlavoie for more insight

You shouldn't need to modify existing code. It should be possible to create a new op, e.g. by copying parts of e.g. `AssignVariableOp`, and instead of using existing classes, define a new class for your resource. So you'd copy `Var` - see `core/framework/resource_var.h` - to create a new `ResourceBase` subclass; its comments have an example. I'm not very familiar with it, but it looks like just a thin wrapper over a pointer to a tensor, and if you make that a `DT_VARIANT` then you can store anything inside those (see `core/kernels/tensor_map.h` for an example).

But before going the route of resources, I think it'd be worth looking into whether you can use regular variants for that. Those aren't references and instead use value semantics, but for things like small containers they can be efficient since you're only really copying pointers. `TensorArray` uses those for example.

**Comment by allenlavoie** [comment]
I think Dan covered it pretty well. Summary writers, Dataset iterators, and tables are all reasonable examples of kernels that handle resource tensors. There isn't anything they can do that isn't public; the C++ custom op API is what they all use to specify their kernels.

Having pointers in the custom op guide sounds reasonable. And probably also a discussion of the issues Dan mentions: stateful things are difficult to handle with gradients and other transformations, so keeping things stateless and conforming to the dataflow model are generally good where possible.

**Comment by scarrazza** [comment]
@mdaley and @allenlavoie thank you for the details. Would be great to have a summary in the docs and/or in [custom-op](https://github.com/tensorflow/custom-op), with an example showing how to construct custom operators for variants as input tensors.

**Comment by saxenasaurabh** [comment]
Additional 2c, `DT_VARIANT` is effectively `void*`. You only need it if you want to have a tensor backed by an arbitrary C++ object. If you all you care about is rectangular tensors then maybe [buffer forwarding](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/scatter_nd_op.cc;l=204-205;drc=3cbb50768909c585d33e99ba10172d1c44c04d6f) is all you need?

**Comment by scarrazza** [comment]
@saxenasaurabh @allenlavoie @mdanatg thank you for the suggestions and references. Let me just share a couple of comments about my current application layout in order to understand if the approaches you are proposing are the best for my specific test case. 

I have to deal with custom operations of large tensors with usually `size=2^35` tf.complex128 elements. So far, in order to avoid copies, I am updating in place the elements stored inside `Tensor::flat<T>::data()` and this obviously disables/destroys useful features such as gradients and `tf.function`. Thus I was thinking about passing variables to custom operators, however at this point, which approach do you suggest for my specific test case? If I understand correctly, buffer forwarding does not guarantees a copy free operator, is the `TensorArray` approach the best so far?

**Comment by allenlavoie** [comment]
So the idea with TensorArray is that you'd have N chunks of size M (N*M ~= 2^35), and only a small number of the N chunks would be updated? If so it sounds plausible that it could work; every time backprop keeps a "copy" of the TensorList it'll make an O(N) copy of the list of chunks, but the data in the chunks themselves should be shared (since they're refcounted Tensors).

Even if TensorList doesn't quite fit (e.g. because the update pattern is sparse but isn't nicely chunked along one axis) I think using similar variant-dtype tensor scheme sounds reasonable and doable with custom ops today. If you want backprop to work you need to provide the illusion that forward pass inputs are available to the backward pass, but you can control how those inputs are stored (dense vs. dense + sparse delta); dense buffer forwarding won't really help since backprop references the old buffers.

**Comment by rmothukuru** [comment]
@scarrazza,
Can you please respond to the above comment? Thanks!
